index.html

```diff
@@ -1,390 +1,1451 @@
+        class LineBreakTransformer {
+          constructor() {
+            this.container = '';
+          }
+
+          transform(chunk, controller) {
+            this.container += chunk;
+            const lines = this.container.split('\r\n');
+            this.container = lines.pop();
+            lines.forEach(line => controller.enqueue(line));
+          }
+
+          flush(controller) {
+            controller.enqueue(this.container);
+          }
+        }
+
+        const lineReader = port.readable
+            .pipeThrough(new TextDecoderStream())
+            .pipeThrough(new TransformStream(new LineBreakTransformer()))
+            .getReader();
+      </pre>
+
+      Some other ways of encoding message boundaries are to prefix each message
+      with its length or to wait a defined length of time before transmitting
+      the next message. Implementing a {{TransformStream}} for these types of
+      message boundaries is left as an exercise for the reader.
+    </aside>
+
+    The {{SerialPort/readable}} getter steps are:
+
+    <ol>
+      <li>
+        If [=this=].{{[[readable]]}} is not `null`, return
+        [=this=].{{[[readable]]}}.
+      <li>If [=this=].{{[[state]]}} is not `"opened"`, return `null`.
+      <li>If [=this=].{{[[readFatal]]}} is `true`, return `null`.
+      <li>Let |stream| be a [=new=] {{ReadableStream}}.
+      <li>
+        Let |pullAlgorithm| be the following steps, given
+        |controller:ReadableStreamDefaultController|:
+        <ol>
+          <li>Let |promise:Promise| be [=a new promise=].
+          <li>[=Resolve=] |promise| with `undefined`.
+          <li>
+            Return |promise| and run the following steps [=in parallel=]:
+            <ol>
+              <li>
+                Let |desiredSize| be
+                |controller|.{{ReadableStreamDefaultController/desiredSize}}
+              <li>
+                Let |buffer| be a newly created {{ArrayBuffer}} of |desiredSize|
+                bytes.
+              <li>
+                Invoke the operating system to read up to |desiredSize| bytes
+                from the port into |buffer|.
+              <li>
+                If no errors were encountered run the following steps:
+                <ol>
+                  <li>Let |bytesRead| be the number of bytes written to
+                  |buffer|.
+                  <li>Let |chunk| be a newly created {{Uint8Array}} over the
+                  slice of |buffer| from `0` to |bytesRead|.
+                  <li>[=Queue a task=] to invoke
+                  |controller|.{{ReadableStreamDefaultController/enqueue()}}
+                  with |chunk| on the [=serial port task source=].
+                </ol>
+              <li>
+                If a buffer overrun condition was encountered, [=queue a task=] to invoke
+                |controller|.{{ReadableStreamDefaultController/error()}} with a
+                <dfn>`BufferOverrunError`</dfn> on the [=serial port task
```

Comment:
Are you defining a new DOMException type? That requires coordinating with the Web IDL spec...

Reply:
I need to define some new errors to express the conditions encountered while receiving data. These could be something other than a `DOMException`. Suggestions?

Reply:
DOMException is probably the right choice. I think we should just open a PR to Web IDL (I guess it's unlikely to be merged, sadly) to coordinate. And maybe you'll get suggestions for existing DOMException types that overlap your proposed new ones, which you could adopt instead.

We haven't really done this before, but IMO Web IDL should be accepting of new DOMException types. The names are just strings; it's good to have a lightweight "registry" for those strings in the Web IDL spec, but it shouldn't be a big deal.

Reply:
I've filed #108 to track this as a follow-up.